nifi-nar-bundles/nifi-windows-event-log-bundle/nifi-windows-event-log-processors/src/main/java/org/apache/nifi/processors/windows/event/log/ConsumeWindowsEventLog.java

```diff
@@ -0,0 +1,282 @@
+            .build();
+
+    public static final List<PropertyDescriptor> PROPERTY_DESCRIPTORS = Collections.unmodifiableList(Arrays.asList(CHANNEL, QUERY, MAX_BUFFER_SIZE, MAX_EVENT_QUEUE_SIZE));
+
+    public static final Relationship REL_SUCCESS = new Relationship.Builder()
+            .name("success")
+            .description("Relationship for successfully consumed events.")
+            .build();
+
+    public static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(Arrays.asList(REL_SUCCESS)));
+    public static final String APPLICATION_XML = "application/xml";
+    private final WEvtApi wEvtApi;
+    private final Kernel32 kernel32;
+    private final ErrorLookup errorLookup;
+    private final String name;
+
+    private Throwable wEvtApiError = null;
+    private Throwable kernel32Error = null;
+
+    private BlockingQueue<String> renderedXMLs;
+    private WEvtApi.EVT_SUBSCRIBE_CALLBACK evtSubscribeCallback;
+    private WinNT.HANDLE subscriptionHandle;
+    private ProcessSessionFactory sessionFactory;
+    private String provenanceUri;
+
+    /**
+     * Framework constructor
+     */
+    public ConsumeWindowsEventLog() {
+        this(null, null);
+    }
+
+    /**
+     * Constructor that allows injection of JNA interfaces
+     *
+     * @param wEvtApi  event api interface
+     * @param kernel32 kernel interface
+     */
+    public ConsumeWindowsEventLog(WEvtApi wEvtApi, Kernel32 kernel32) {
+        this.wEvtApi = wEvtApi == null ? loadWEvtApi() : wEvtApi;
+        this.kernel32 = kernel32 == null ? loadKernel32() : kernel32;
+        this.errorLookup = new ErrorLookup(this.kernel32);
+        if (kernel32 != null) {
+            name = Kernel32Util.getComputerName();
+        } else {
+            // Won't be able to use the processor anyway because native libraries didn't load
+            name = null;
+        }
+    }
+
+    private WEvtApi loadWEvtApi() {
+        try {
+            return WEvtApi.INSTANCE;
+        } catch (Throwable e) {
+            wEvtApiError = e;
+            return null;
+        }
+    }
+
+    private Kernel32 loadKernel32() {
+        try {
+            return Kernel32.INSTANCE;
+        } catch (Throwable e) {
+            kernel32Error = e;
+            return null;
+        }
+    }
+
+    /**
+     * Register subscriber via native call
+     *
+     * @param context the process context
+     */
+    @OnScheduled
+    public void onScheduled(ProcessContext context) throws Exception {
+        String channel = context.getProperty(CHANNEL).getValue();
+        String query = context.getProperty(QUERY).getValue();
+
+        renderedXMLs = new LinkedBlockingQueue<>(context.getProperty(MAX_EVENT_QUEUE_SIZE).asInteger());
+        provenanceUri = new URI("winlog", name, "/" + channel, query, null).toASCIIString();
+
+        evtSubscribeCallback = new EventSubscribeXmlRenderingCallback(getLogger(), s -> {
+            try {
+                renderedXMLs.put(s);
```

Comment:
This should be changed to either "add" or "offer". In the event there is a large number of Windows Events that get created in a short span (in relation to the scheduling period) this could create an endless number of threads waiting for space in the queue and bring down the instance. 

With "add" or "offer" you can handle it how you see appropriate for this processor (is there any way to ask Windows event log to try again later?). If necessary/appropriate throw an exception and/or log the error. While this would lead to data loss, that is better than taking the whole instance down (leading to much more data loss).